### ISSUE TITLE
APIMockHelper assumes first service will have an endpoint for some arbitrary tenant

### ISSUE THREAD

**Issue by lvh** [rackerlabs/mimic#228]
```python
        tenant_id = self.auth.service_catalog_json["access"]["token"]["tenant"]["id"]
        service_name = apis[0].catalog_entries(tenant_id)[0].name
        self.uri = self.get_service_endpoint(service_name)
```

First off, I'm not sure why this is taking the first API/catalog entry. I guess it assumes that there will only ever be exactly one API under test, but then I'm not sure why you instantiate it with a collection of APIs.

Secondly, this is just for some arbitrary test tenant. I'm writing Identity Admin; it definitely shouldn't show up for (most) tenants. Is there some policy in place for distinguishing between admin and non-admin tenants based on their tenant id?

Perhaps the resolution for this particular ticket is just to put a `try/except` there and set `self.uri = None`?

**Comment by glyph** [comment]
Please feel free to refactor this; it exists in this state because there are a bunch of tests which assume that this one random tenant will be selected. Perhaps there's a nice way to change it without breaking all the tests, in which case, please do so.

Do you think this is sufficiently actionable to remain an open issue, though?  Should we just put a comment in the code explaining the intended direction of this (i.e. each test should ask specifically, nobody should rely on `self.uri`)?

**Comment by lvh** [comment]
Well, yes and no. A comment might help explain why it is addressed, but some code change needs to happen in order to make plugins that don't expose things to arbitrary tenants can use this code.

(Does that mean that there are currently no standards in mimic for designating e.g. an admin tenant? Can I just pick one and get cracking?)

**Comment by glyph** [comment]
The "special"-ness of a tenant is just per-tenant state, so just put things onto the session.  I think this *might* mean you need to modify `IAPIMock.catalog_entries` to take a `SessionStore` as well as a `tenant_id`...

**Comment by lvh** [comment]
Why should that get passed to `catalog_entires` and not just passed to `APIMock` instances that need it? Also, I've found some stuff right now in `canned_responses`, but that seems to suggest that literally every tenant is just an identity admin/service admin?

Also, are sessions scoped per tenant, or per (tenant, service) tuple? Roles seem like they should be per-tenant.

**Comment by glyph** [comment]
> Why should that get passed to `catalog_entires` and not just passed to `APIMock` instances that need it?

I'm not sure what you mean; you can't pass stuff to "an instance", you pass it to a method :).  `IAPIMock` objects (at least in a plugin context) are constructed at module level, and only find out about their `MimicCore` later; although presently I don't think they find out about it *directly* at all; the `session_store` is simply passed to `IAPIMock.resource_for_region`.

> Also, I've found some stuff right now in `canned_responses`, but that seems to suggest that literally every tenant is just an identity admin/service admin?

Eh... yeah, I guess so.  We should probably start somehow segregating this stuff off, because while we don't want Mimic to be authenticating too much, we also probably don't want real clients to all start relying on being identity admin to work.

**Comment by lvh** [comment]
> > Why should that get passed to `catalog_entires` and not just passed to `APIMock` instances that need it?

> I'm not sure what you mean; you can't pass stuff to "an instance", you pass it to a method :).  `IAPIMock` objects (at least in a plugin context) are constructed at module level, and only find out about their `MimicCore` later; although presently I don't think they find out about it *directly* at all; the `session_store` is simply passed to `IAPIMock.resource_for_region`.

I guess `__init__` is a method, yes; I meant passed at creation.

>> Also, I've found some stuff right now in `canned_responses`, but that seems to suggest that literally every tenant is just an identity admin/service admin?

> Eh... yeah, I guess so.  We should probably start somehow segregating this stuff off, because while we don't want Mimic to be authenticating too much, we also probably don't want real clients to all start relying on being identity admin to work.

OK, cool. So, next ticket: catalog_entries should take a session store?

**Comment by glyph** [comment]
> Also, are sessions scoped per tenant, or per (tenant, service) tuple? Roles seem like they should be per-tenant.

If you are messing with sessions, https://github.com/rackerlabs/mimic/issues/158 is probably interesting and relevant to future plans.

To answer your question more directly though, consider [this representative usage from `NovaApi`](https://github.com/rackerlabs/mimic/blob/master/mimic/rest/nova_api.py#L79-L85) - `SessionStore.session_for_tenant_id` is scoped per tenant, but `.data_for_api` adds the additional scoping to a specific API mock.  ("service" is a bit too vague of a term here so I definitely prefer to use the Mimic-specific word for the particular object it's scoped to.)

**Comment by glyph** [comment]
> OK, cool. So, next ticket: catalog_entries should take a session store?

Yes, that seems like the most straightforward extension of the current API to achieve this.

**Comment by glyph** [comment]
> Roles seem like they should be per-tenant.

Sorry, I didn't really cover this in my previous response.  Yes, they should be.  But giving the "role" attribute a specific semantic meaning in context needs to be done in the context of a particular namespace, and that namespace is defined by a service (and its accompanying session-data-creator thing, which is why I referenced #158 ).  Currently it's just a dictionary key, and when #158 expands its interface to be able to create session data we could accommodate this too: we should have a stand-in Identity API mock which is used for this key, so that when a service wants to manipulate Identity data on a given session, it can do so in much the same way that the control-plane plugin for nova manipulates session data for servers; the service-catalog-generation code can call:

```python
from mimic.model.identity import identity_pseudo_mock
# ...
identity_session = store.session_for_tenant_id(tenant_id).data_for_api(identity_pseudo_mock)
is_admin = identity_session.am_i_admin_y_or_n()
```

and so on.

Does that make sense?

**Comment by lvh** [comment]
Uh, sure.

**Comment by glyph** [comment]
The addition of the "Uh" makes the "sure" sound like "no" :smile: . Anything I can expand on?

**Comment by lvh** [comment]
Nope; I think that makes enough sense that I can follow it, but those sequence of steps definitely weren't obvious to me as the correct thing to do :)

**Comment by lvh** [comment]
One of the things that is a little confusing is that that nova_api use uses an actual API, but your example seems to suggest a sentinel value whose only job is to denote "identity stuff".
